Updating Vue Quickstart for auth0-spa-js

stevehobbsdev opened the pull request:
This PR updates the previous Vue SPA quickstart with new guidance that uses the auth0-spa-js JavaScript client.

Samples PR here: https://github.com/auth0-samples/auth0-vue-samples/pull/82

Commit:
Rewrite of 01-Login using auth0-spa-js

Commit:
Minor tweaks to 01-Login

- Modified heading to be more specific to installing SDK
- Added a section on running the app with the correct port
- Removed audience from config
- Added a note about putting auth values into source control

Commit:
Reworded 'service' to 'wrapper'

Commit:
Updated tutorial 03 for auth0-spa-js

Commit:
Fixed up typos, and minor tweaks

Commit:
Renamed 03-Calling-an-API to 02-Calling-an-API

This is inkeeping with other new spa-js samples.

erjohnson commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+
+<!-- markdownlint-disable MD002 MD041 -->
+
+Most single-page apps use resources from data APIs. You may want to restrict access to those resources so that only authenticated users with sufficient privileges can access them. Auth0 lets you manage access to these resources using [API Authorization](/api-auth).
```
Most single-page apps use resources from data APIs. You may want to restrict these resources so only authenticated users with sufficient privileges can access them. Auth0 lets you manage access to these resources using [API Authorization](/api-auth).

erjohnson commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+
+Most single-page apps use resources from data APIs. You may want to restrict access to those resources so that only authenticated users with sufficient privileges can access them. Auth0 lets you manage access to these resources using [API Authorization](/api-auth).
+
+This tutorial shows you how to create a simple API using [Express](https://expressjs.com) that validates incoming JSON Web Tokens. You will then see how to call this API using an access token granted by the Auth0 authorization server.
```
In this tutorial, you'll create a simple API using [Express](https://expressjs.com) to validate incoming JSON Web Tokens. You will then see how to call this API using an access token granted by the Auth0 authorization server.

erjohnson commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+
+## Add the Audience
+
+The API backend that you created above expects API calls to have a JSON Web Token attached so that it can authorize the call. To do this, you must instruct Auth0 to emit an access token in JWT format. This is done by supplying the API identifier when your users authenticate, telling Auth0 which API you want an access token for.
```
The API backend you created above expects API calls to have a JSON Web Token for authorization.

erjohnson commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+
+## Call the API Using an Access Token
+
+The frontend Vue.js application should be modified to include a page that calls the API using an access token. Similar to the previous tutorial, this includes modifying the Vue router and adding a new view with a button that calls the API.
```
Modify the frontend Vue.js application to include a page that calls the API using an access token. Like the previous tutorial, you'll need to change the Vue router and add a new view with a button that calls the API.

Commit:
Applied changes based on content review

joshcanhelp commented on articles/quickstart/spa/_includes/_calling_api_create_backend.md:
```diff
@@ -2,7 +2,7 @@
 ## Create the Backend API
 
-For this example, you'll create an [Express](https://expressjs.com/) server that acts as the backend API. This API will expose an endpoint to validate incoming ID Tokens before returning a response.
+For this example, you'll create an [Express](https://expressjs.com/) server that acts as the backend API. This API will expose an endpoint to validate incoming [JSON Web Tokens](https://auth0.com/docs/jwt) before returning a response.
```
"JSON Web Tokens" is a little too general in this case. They are JWTs but, specifically, they are access tokens. Maybe "JWT-formatted access tokens" or similar?

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+useCase: quickstart
+---
+
+<!-- markdownlint-disable MD002 MD041 -->
```
TIL! 🎉

stevehobbsdev replied:
:D

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+
+Most single-page apps use resources from data APIs. You may want to restrict these resources so only authenticated users with sufficient privileges can access them. Auth0 lets you manage access to these resources using [API Authorization](/api-auth).
+
+In this tutorial, you'll create a simple API using [Express](https://expressjs.com) to validate incoming JSON Web Tokens. You will then see how to call this API using an access token granted by the Auth0 authorization server.
```
Same feedback on "JSON Web Tokens"

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+
+### Set up a proxy to the backend API
+
+To call the API from the frontend application, the development server must be configured to proxy requests through to the backend API. To do this, add a `vue.config.js` file to the root of the project and populate it with the following code:
```
Is this essential to the quickstart?

stevehobbsdev replied:
@joshcanhelp It's a tricky one that I have constantly battled with. It's not essential to showing how Auth0 works, no. If we take this out, we'd probably have to take out everything to do with creating the backend server.

I'd be in favour of doing this as it would condense the quickstart down further to just the essentials while still showing how to _call_ an API. We could leave the actual backend in the samples as an example for the reader to view.

The only issue then is that the other SPA quickstarts have this backend. Options:

1) Leave the backend documentation where it is for consistency with other SPA quickstarts
2) Remove it in favour of making the quickstart more concise

Personally I'm leaning towards 2. We could try it and see what kind of feedback we get on it, if any?

joshcanhelp replied:
Now I'm split :)

My default decision is to remove anything not related to the task at hand. If it does not improve the experience, increase learning, or reduce overall complexity, I don't think it should be there. If, however, this increases the chance of success for some reason (smooth out gaps in a specific platform or something), then that's a positive. 

I'll leave it up to you but it _seems_ like this just makes a nice URL for the API which, in this case, is not relevant for the task at hand.

stevehobbsdev replied:
It's more to make the API accessible on the same origin. When considering the bigger picture, it makes configuration and deployment a bit easier too.

I think it's more a question of whether the docs needs to detail how the sample app is built, or leave it out of the docs and have readers study the sample app for this kind of detail.

For now, I'm going to leave it as it is, as this is detail is also in the other SPA quickstarts and we haven't had any feedback on it to make us think it's detrimental to the reader's experieance. Howver, I do hope that this is something that the current work into quickstart principals can help definitively answer.

joshcanhelp replied:
> It's more to make the API accessible on the same origin.

Ahhhh, that didn't occur to me. Makes sense to me. Thank you for the thorough explanation.

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
````diff
@@ -0,0 +1,234 @@
+
+If you are already running the application using `npm run serve`, then stop that process for now. Restart the app using the following command, which will run both the frontend app and the backend API together:
+
+```bash
````
Nice little command here!

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+
+## Add the Audience
+
+The API backend you created above expects API calls to have a JSON Web Token for authorization. To do this, you must instruct Auth0 to emit an access token in JWT format. This is done by supplying the API identifier when your users authenticate, telling Auth0 which API you want an access token for.
```
```suggestion
The API backend you created above expects API calls to have a JSON Web Token for authorization. To do this, you must instruct Auth0 to issue an access token in JWT format. This is done by supplying the API identifier in an `audience` parameter when users authenticate, telling Auth0 which API you want an access token for.
```

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
```diff
@@ -0,0 +1,234 @@
+// .. other imports
+
+// NEW - import the audience
+import { domain, clientId, audience } from "../auth_config.json";
```
Great way to break the config out!

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
````diff
@@ -0,0 +1,234 @@
+```html
+<template>
+  <div>
+    <button @click="callApi">Ping</button>
````
```suggestion
    <button @click="callApi">Call</button>
```

stevehobbsdev replied:
Will change this on the sample too.

joshcanhelp commented on articles/quickstart/spa/vuejs/02-calling-an-api.md:
````diff
@@ -0,0 +1,234 @@
+```
+
+:::panel Checkpoint
+Now you will be able to run the application, browse to the "External API" page and press the "Ping" button. The application will make a call to the external API endpoint and produce a message on the screen that says "Your Access Token was successfully validated!".
````
```suggestion
Now you will be able to run the application, browse to the "External API" page and press the "Call" button. The application will make a call to the external API endpoint and produce a message on the screen that says "Your Access Token was successfully validated!".
```

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
 <%= include('../../_includes/_login_preamble', { library: 'Vue.js' }) %>
 
-### Create an Authentication Service
+## Create a Sample Application
```
It's very confusing that a "Integrate Auth0 in your Application" section appears above this with instructions about importing out of the blue. Then this section also gets into installing. I would remove the first one.

stevehobbsdev replied:
I agree, and in fact _loads_ of people have an issue with the "Integrate Auth0..." instructions as we don't tell them at that point which file it should go in. They're intended as generic instructions.

I'm going to remove this include altogether as well as the "Authentication with Auth0" include and see what difference that makes. Can you check out the rendered article again to see what you think?

joshcanhelp replied:
Much better!!

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
-class AuthService extends EventEmitter {
+```bash
+# Install the CLI
+npm install -g @vue/cli
````
Might want to add some kind of symbol showing the command prompt, like `❯` or `$` in bash commands.

stevehobbsdev replied:
Generally we haven't done this previously as it makes the command harder to copy and paste.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
-    });
-  }
+# Create the application using the Vue CLI.
+# When asked to pick a preset, accept the defaults
```
The default is `yarn` on my install ... may want to mention npm (if that matters)

stevehobbsdev replied:
I think you're getting yarn as an option because you've got it installed - the article will work fine with either.

One thing that has come out of the quickstart principals talks (from the few I've been involved in) is that we should default to `npm` when talking about it on the docs and assume that's what the reader will be using. If an option for `yarn` is presented, then it's likely that the reader will probably just use that if they so chose, and would be comfortable translating the commands as they come up.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
 
-export default new AuthService();
+# Add the router, as we will be using it later
+vue add router
```
Asks if I should use history mode

stevehobbsdev replied:
Good point, will mention this.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
+```
 
-Add some additional methods to `AuthService` to fully handle authentication in the app.
+At this point, you can leave the application running in the background, as it will reload whenever you make changes. Note the port number — if you intend your app to run on a different port, remember to adjust this number to suit.
````
```suggestion
At this point, you can leave the application running in the background, as it will reload whenever you make changes. Note the port number — if you intend your app to run on a different port, remember to adjust this number throughout the quickstart.
```

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
-  idToken = null;
-  profile = null;
-  tokenExpiry = null;
+To implement the wrapper, create a new folder called `auth` inside the `src` folder, and then create a new file called `index.js` inside. Populate this file with the following content:
```
I don't have a `src` folder at this point. Create one?

stevehobbsdev replied:
I'm assuming this was because (from your Slack message) that you hadn't switched into the project folder - is that the case?

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
-      appState: customState
-    });
-  }
+:::note
```
Would it be better to maybe put this in a repo or Gist now so it can be copied whole-sale? I would expect all of this code in this tutorial to be understood rather than just boilerplate.

stevehobbsdev replied:
Certainly an option. Right now I've followed what we've done for the React and Angular samples. I'm hoping that it won't be too long before we take these snippets out and move them into their own libraries, in which case we'll come back and update these docs.

I don't have a strong opinion either way, except that the way it is now isn't _hugely_ inhibiting. It keeps the reader on the page (as opposed to redirecting them to a Gist) and there is a single button they can press to copy the whole snippet and paste it in. What do you think?

joshcanhelp replied:
I'm good with that.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
-
-```js
-// src/plugins/auth.js
+The `options` object passed to the plugin is used to provide the values for `clientId` and `domain`. For this example, create a new file `auth_config.json` in the root directory of the application alongside your `package.json` file, and populate it with your tenant values:
````
If I follow the steps in this QS as-is, I don't have a `package.json`, only `package-lock.json` (I didn't run `npm init`).

stevehobbsdev replied:
Strange - again relating to being in the right folder?

joshcanhelp replied:
Yes, sorry, I guess I didn't remove all the comments related to that 🤷‍♂

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
-
-```js
-// src/plugins/auth.js
+The `options` object passed to the plugin is used to provide the values for `clientId` and `domain`. For this example, create a new file `auth_config.json` in the root directory of the application alongside your `package.json` file, and populate it with your tenant values:
````
```suggestion
The `options` object passed to the plugin is used to provide the values for `clientId` and `domain`. For this example, create a new file `auth_config.json` in the root directory of the application alongside your `package-lock.json` file, and populate it with the values from your application created above:
```

stevehobbsdev replied:
Have applied this change locally without the `package.json` bit.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
 ```
 
-This plugin provides access to the `AuthService` class from each component, through the `this.$auth` property. It also provides a mechanism for when the login state changes, for components that implement a `handleLoginEvent` method.
+:::note
````
```suggestion
:::warning
```

joshcanhelp replied:
May also want to note that nothing here is sensitive, just configuration.

stevehobbsdev replied:
Nice, have adjusted the wording here.

stevehobbsdev replied:
Also, applied the change locally.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
+<!-- ![hosted login](/media/articles/web/hosted-login.png) -->
 
-Open `main.js` and install the plugin:
+Finally, open `src/main.js` and use `Vue.use` to install the plugin. Notice that the configuration for your Auth0 app is imported from `auth_config.json` and supplied to the plugin:
```
Need to create this file

joshcanhelp replied:
My bad

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
-
-```js
-<!-- src/components/Callback.vue -->
+Open the `views/Home.vue` file and modify the template to include two buttons that enable the user to log in and log out:
````
Again, need to create this.

joshcanhelp replied:
Again, my bad

joshcanhelp reviewed (changes requested):
Got stuck ... sending feedback over Slack but submitting this for now.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
 import Vue from "vue";
 import App from "./App.vue";
+import router from './router'
+import { domain, clientId } from "../auth_config.json";
```
Maybe a note that this is new as well?

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
+    <img alt="Vue logo" src="../assets/logo.png" />
+    <HelloWorld msg="Welcome to Your Vue.js App" />
+
+    <!-- Check that the SDK client is not currently loading before accessing is methods -->
```
I didn't realize this was the only section added at first ... easy to fix but might be better guidance we can provide?

stevehobbsdev replied:
I wish there was a solution to this.

One thing I tried was using the `diff` type to highlight changes, which gives us this:

<img width="896" alt="image" src="https://user-images.githubusercontent.com/766403/65439197-b2b37000-de1e-11e9-94ca-c6c3ffce1981.png">

But it's still not perfect. In an ideal world, the new additions would be colorized normally as per the language syntax, and everything else would be greyed out.

stevehobbsdev replied:
I wonder if we can provide a custom colour scheme to at least grey-out the non-changes.. 🤔

stevehobbsdev replied:
Something like this:

<img width="896" alt="image" src="https://user-images.githubusercontent.com/766403/65439793-a4b21f00-de1f-11e9-9a42-ea89d1b41150.png">

joshcanhelp replied:
I would **LOVE** to have that option. The upcoming ID labs would really use that as well. Not sure if that's possible with the code highlighting library we're using. Maybe something that @davidpatrick could help with?

stevehobbsdev replied:
The above screenshot ☝️ is easy to do with a bit of styling. Nothing else in the docs project uses the `diff` language type so it would be easy to target.

What's not so easy is having the diff show up in normal syntax colours and grey out everything else. Looking around, there's a few SO questions and blog posts about a similar thing but nothing concrete.

I'm going to put aside some time later to try and work out a possible solution.

joshcanhelp replied:
We can put this aside for now since it should not block this PR. Let me know if I can help at all.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
+
+```js
 <script>
+// .. imports removed for brevity
````
Not a lot of brevity achieved 😆

stevehobbsdev replied:
I thought something was better than nothing 😄

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
-import Router from 'vue-router';
-import Callback from './components/Callback';
+:::panel Checkpoint
+Try running the application at this point. Provided you have configured your Auth0 credentials correctly, you should be redirected to the Auth0 Universal Login page, and back to your app against once you have logged in.
```
The callback mentioned at the top of the page - `http://localhost:3000/callback` - apparently is not correct. The one used is `http://localhost:3000` so I got a callback error when I ran it.

stevehobbsdev replied:
Good catch, yes the `/callback` should now be removed.

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
-<%= include('../../_includes/_hosted_login_customization') %>
-
-When the application first starts up, a call to `renewTokens` is made that tries to reinitialize the user's login session, if it is detected that they should already be logged in. This would be the case, for example, if the user logged in and then refreshed the browser window.
+Notice that the configuration file created earlier has been imported and used to initialize the plugin. A custom redirect callback has also been supplied, which redirects the user to a protected route after they have authenticated (this is covered a bit later).
```
> A custom redirect callback has also been supplied ...

Has it? See comment below RE: callback error.

stevehobbsdev replied:
From a previous comment, I've added in a comment above the import which *should* make it more visible - what do you think?

joshcanhelp replied:
I think my confusion was around the wording:

> custom redirect callback ...

You're talking about a redirect after logging in here, correct? The redirect to the profile page works fine but I think "callback" used in this context is a little misleading (we typically use "callback" to indicate where the auth is processed which, in this case, is just the homepage).

stevehobbsdev replied:
I see what you're saying. I've changed the wording to "handler" - how does that sound?

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
```diff
@@ -2,475 +2,441 @@
 // src/router.js
 
 //.. other imports
 import Profile from "./views/Profile.vue";
```
Missed this at first ...

stevehobbsdev replied:
Reworded slightly and added comments

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
 
 ```js
-// src/router.js
+import { getInstance } from "./authService";
````
Getting an error with this:

```
Failed to compile.

./src/auth/authGuard.js
Module not found: Error: Can't resolve 'authService' in '/Users/josh-cunningham/Sites/quickstarts/spa-qs-vue/src/auth'
```

stevehobbsdev replied:
Ah, the import was wrong. I've fixed it now

joshcanhelp replied:
The change:

```js
import { getInstance } from "./auth";
```

... still gave me a compile error. I changed to:

```js
import { getInstance } from "./index";
```

... and that worked.

stevehobbsdev replied:
Ah of course, because you're in a file in the same folder. Thanks for the correction!

joshcanhelp commented on articles/quickstart/spa/vuejs/_includes/_centralized_login.md:
````diff
@@ -2,475 +2,441 @@
+```
+
+:::panel Checkpoint
+Start the application and make sure you are logged out. Now, modify the URL in your browser's navigation bar to include the `/profile` path on the end. This should cause the application to ask you to authentication. Once you have logged in, the profile page should be visible without you having to navigate there manually.
````
Could not reach this, see comment above about compile error.

stevehobbsdev replied:
Should be ok with the above fixed.

joshcanhelp reviewed (changes requested):
Super-slick, nice work once again!

Two things to note generally:

- I would be consistent with the path/fileName as a comment on top of the code blocks. I think that's a helpful addition.
- I don't have a great solution for you but I was caught up a few times with missing a new line of code to add. I would be very consistent with how you call those out. I wish we could grey out code but I typically add a consistent comment line above like `// Add the line below 👇` for a single line or `// Add the code below 👇` for a block or comment out the replaced code with a note and add a line above the added block.

Commit:
Removed preamble includes for conciseness

- Removed spa-js install instructions - they are includede further down
the article
- Added a flag into the 'getting_started' include to toggle off the
install instructions
- Corrected JSON Web Tokens wording for clarity

Commit:
Apply suggestions from code review

Co-Authored-By: Josh Cunningham <josh@joshcanhelp.com>

Commit:
Added a comment about using history mode when adding router

Commit:
Apply suggestions from code review

Co-Authored-By: Josh Cunningham <josh@joshcanhelp.com>

Commit:
Applied suggestions from review locally

Commit:
Minor fixes from review

Removed /callback from the callback URL

Fixed an issue with the authGuard import

Commit:
Minor amendments based on feedback

joshcanhelp reviewed (approved):
We are good here :) Very nice work on this 👍
